### PR TITLE
[Metrics] Confusion matrix class interface

### PR DESCRIPTION
**Pull request by SkafteNicki**
## What does this PR do?

* Adds back confusion matrix class interface
* Unify with functional interface
* Updated old test to new interface

## Before submitting
- [ ] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests? 
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## PR review
 - [ ] Is this pull request ready for review? (if not, please submit in draft mode)

Anyone in the community is free to review the PR once the tests have passed.    
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/metrics/classification/confusion_matrix.py [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+        Update state with predictions and targets.
+
+        Args:
+            preds: Predictions from model
+            target: Ground truth values
+        """
+        confmat = _confusion_matrix_update(preds, target, self.num_classes, self.threshold)
+        self.confmat += confmat
+
+    def compute(self):
```
Comment (justusschock) [inline pytorch_lightning/metrics/classification/confusion_matrix.py:104]:
```suggestion
    def compute(self) -> torch.Tensor:
```

**Review by justusschock — APPROVED** [review]

File: pytorch_lightning/metrics/classification/confusion_matrix.py (hunk 1 above)
Comment (SkafteNicki) [reply pytorch_lightning/metrics/classification/confusion_matrix.py:104]: Regardless how much I try to remember to include typing, I always forget it somewhere. Thanks :]

**Comment by s-rog** [comment]
The confusion matrix is computed on each GPU and reduced (summed) across the group right? I Just wanna make sure I'm understanding the code correctly!

**Comment by SkafteNicki** [comment]
@s-rog, yes whenever `.compute` method is called, this function is called:
https://github.com/PyTorchLightning/pytorch-lightning/blob/b26c71eadf11d3a6aa9428504b8e646e0a8a03f8/pytorch_lightning/metrics/metric.py#L175-L193
which first do a `gather_all` per state variable (those added through `self.add_state`) and then apply whatever function is specified in `self.add_state`.

**Review by s-rog — APPROVED** [review]
Thanks for the clarification!

**Review by tchaton — APPROVED** [review]
Great PR !

**Review by rohitgr7 — APPROVED** [review]

File: pytorch_lightning/metrics/classification/confusion_matrix.py [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+    This is the case for binary and multi-label logits.
+
+    If preds has an extra dimension as in the case of multi-class scores we perform an argmax on ``dim=1``.
+
+    Args:
+        num_classes: Number of classes in the dataset.
+        normalize: normalization mode for confusion matrix. Default is None,
+            meaning no normalization. Choose between normalization over the
+            targets (`'true`', most commonly used), the predictions (`'pred`') or
+            over hole matrix (`'all'`)
```
Comment (rohitgr7) [inline pytorch_lightning/metrics/classification/confusion_matrix.py:46]: maybe put them as bullet points?

**Comment by Vichoko** [comment]
Nice PR! I need this. *o*

File: pytorch_lightning/metrics/classification/confusion_matrix.py [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+        num_classes: Number of classes in the dataset.
+        normalize: Normalization mode for confusion matrix. Choose from
+
+            * `None`, no normalization (default)
+
+            * `'true'`, normalization over the targets (most commonly used)
+
+            * `'pred'`, normalization over the predictions
+
+            * `'all'`, normalization over the hole matrix
```
Comment (rohitgr7) [inline pytorch_lightning/metrics/classification/confusion_matrix.py:51]:
```suggestion
            - ``None``: no normalization (default)
            - ``'true'``: normalization over the targets (most commonly used)
            - ``'pred'``: normalization over the predictions
            - ``'all'``: normalization over the whole matrix
```

**Comment by Vichoko** [comment]
In which ptl version this will be released?

**Comment by ydcjeff** [comment]
> In which ptl version this will be released?

it could be in coming point release on Tuesday.

**Comment by NumesSanguis** [comment]
According to the [release changelog](https://github.com/PyTorchLightning/pytorch-lightning/releases/tag/1.0.5), ConfusionMatrix should be in PL v1.0.5.
However, trying to import it, fails:

```python
>>> from pytorch_lightning.metrics import ConfusionMatrix
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ImportError: cannot import name 'ConfusionMatrix' from 'pytorch_lightning.metrics' (*path*/python3.8/site-packages/pytorch_lightning/metrics/__init__.py)
```

Colab: https://colab.research.google.com/drive/1QHFCBtGEv215DuLw6Suqi9A1WHiExlZR?usp=sharing

**Comment by ydcjeff** [comment]
@NumesSanguis I think they cherry-picked the commits for 1.0.5 release so it didn't include in it. maybe in 1.1

**Comment by NumesSanguis** [comment]
@ydcjeff Then the patch notes are wrong:

> Added ConfusionMatrix class interface (#4348)
> Added multiclass AUROC metric (#4236)

You can find this Metric also already in the docs:
https://pytorch-lightning.readthedocs.io/en/latest/metrics.html#confusionmatrix

**Comment by ydcjeff** [comment]
Yep, we are in the transition of point release and minor release, we will edit the changelog

**Comment by s-rog** [comment]
https://github.com/PyTorchLightning/pytorch-lightning/pull/4505/

**Comment by NumesSanguis** [comment]
I don't see an associated Feature Request, so I hope bug reports here are ok.

@SkafteNicki I tried this ConfusionMatrix using the master branch (`pip install git+https://github.com/PytorchLightning/pytorch-lightning.git@master --upgrade`), but there is an error when the last batch size does not match earlier batches:
```python
     25                              threshold: float = 0.5) -> torch.Tensor:
     26     preds, target = _input_format_classification(preds, target, threshold)
---> 27     unique_mapping = (target.view(-1) * num_classes + preds.view(-1)).to(torch.long)
     28     bins = torch.bincount(unique_mapping, minlength=num_classes ** 2)
     29     confmat = bins.reshape(num_classes, num_classes)

RuntimeError: The size of tensor a (32) must match the size of tensor b (16) at non-singleton dimension 0
```

Colab: https://colab.research.google.com/drive/1QHFCBtGEv215DuLw6Suqi9A1WHiExlZR?usp=sharing

**Comment by NumesSanguis** [comment]
@SkafteNicki There is also an issue (if you make sure the above batch size does not show up) when trying to log the Confusion Matrix with the v1.0.x logging approach:

```python
self.confmat(output.softmax(dim=1), labels)
self.log(f"confmat/val", self.confmat)
```

Gives:
```python
/usr/local/lib/python3.6/dist-packages/pytorch_lightning/trainer/logging.py in metrics_to_scalars(self, metrics)
     46         for k, v in metrics.items():
     47             if isinstance(v, torch.Tensor):
---> 48                 v = v.item()
     49 
     50             if isinstance(v, dict):

ValueError: only one element tensors can be converted to Python scalars
```

Colab: https://colab.research.google.com/drive/1GRi0VhRTbDGdj4D1HUMo49Pe7__tWe31?usp=sharing

**Comment by s-rog** [comment]
@NumesSanguis Mind opening a separate issue?

**Comment by SkafteNicki** [comment]
> 
> 
> I don't see an associated Feature Request, so I hope bug reports here are ok.
> 
> @SkafteNicki I tried this ConfusionMatrix using the master branch (`pip install git+https://github.com/PytorchLightning/pytorch-lightning.git@master --upgrade`), but there is an error when the last batch size does not match earlier batches:
> 
> ```python
>      25                              threshold: float = 0.5) -> torch.Tensor:
>      26     preds, target = _input_format_classification(preds, target, threshold)
> ---> 27     unique_mapping = (target.view(-1) * num_classes + preds.view(-1)).to(torch.long)
>      28     bins = torch.bincount(unique_mapping, minlength=num_classes ** 2)
>      29     confmat = bins.reshape(num_classes, num_classes)
> 
> RuntimeError: The size of tensor a (32) must match the size of tensor b (16) at non-singleton dimension 0
> ```
> 
> Colab: https://colab.research.google.com/drive/1QHFCBtGEv215DuLw6Suqi9A1WHiExlZR?usp=sharing

This example does not reflect a real example. When you are creating artificial labels, when you do `batch[1].sigmoid()` you are creating a label tensor along the *feature* dimension, but should instead create it along the *batch dimension* like this i.e. `batch[:,0].sigmoid()`.

> 
> 
> @SkafteNicki There is also an issue (if you make sure the above batch size does not show up) when trying to log the Confusion Matrix with the v1.0.x logging approach:
> 
> ```python
> self.confmat(output.softmax(dim=1), labels)
> self.log(f"confmat/val", self.confmat)
> ```
> 
> Gives:
> 
> ```python
> /usr/local/lib/python3.6/dist-packages/pytorch_lightning/trainer/logging.py in metrics_to_scalars(self, metrics)
>      46         for k, v in metrics.items():
>      47             if isinstance(v, torch.Tensor):
> ---> 48                 v = v.item()
>      49 
>      50             if isinstance(v, dict):
> 
> ValueError: only one element tensors can be converted to Python scalars
> ```
> 
> Colab: https://colab.research.google.com/drive/1GRi0VhRTbDGdj4D1HUMo49Pe7__tWe31?usp=sharing

`self.log` does not allow logging anything else than scalar tensors. This is a general limitation of `self.log` (not specific related to confusion matrix.

**Comment by Vichoko** [comment]
If there is no logging support through `self.log`, how I'm supposed to aggregate the confusion matrix across batches and GPUs (on a multi-GPU scheme)?

**Comment by SkafteNicki** [comment]
@Vichoko depending on what logger you are using, you can create the confusion matrix as a figure and log it. For example using the default tensorboard logger:
``` python
import matplotlib.pyplot as plt
...
def training_step(self, batch, batch_idx):
    ...
    step_confmat = self.metric(pred, target)
    fig = plt.figure(); plt.imshow(step_confmat)
    self.logger.experiment.add_figure('step_confmat', fig, global_step=self.global_step)

def training_epoch_end(self, outputs):
    epoch_confmat = self.metric.compute()
    fig = plt.figure(); plt.imshow(epoch_confmat)
    self.logger.experiment.add_figure('epoch_confmat', fig, global_step=self.global_step)
```

**Comment by NumesSanguis** [comment]
> This example does not reflect a real example. When you are creating artificial labels, when you do batch[1].sigmoid() you are creating a label tensor along the feature dimension, but should instead create it along the batch dimension like this i.e. batch[:,0].sigmoid().

Sorry, somehow I had my own classification problem in mind where I assumed that `batch` was a tuple of `(inputs, labels)`, hence I used `batch[1]`.

-----
Thank you for the confusion matrix example, that was actually what I was searching for!

**Comment by NumesSanguis** [comment]
I assume the x-axis (bottom) to represent the *Predicted label*, and the y-axis (left) to be the *True label*, but could this be added to the documentation?

**Comment by SkafteNicki** [comment]
We follow sklearns way of representing confusion matrices (https://scikit-learn.org/stable/modules/generated/sklearn.metrics.confusion_matrix.html#sklearn.metrics.confusion_matrix) which indeed is predicted by x-axis/columns and true by y-axis/rows.
@NumesSanguis please feel free to open a PR with recommended doc update.

**Comment by Borda** [comment]
@SkafteNicki @NumesSanguis it shall be fixed in #4505

**Comment by Vichoko** [comment]
Changelog implied that this would be released on 1.0.6 but it wasn't :c

**Pull request merged**